modules/runner/src/main/java/org/apache/ignite/internal/app/IgniteImpl.java

```diff
@@ -568,4 +628,44 @@ private static Path getPartitionsStorePath(Path workDir) {
 
         STOPPING
     }
+
+    /**
+     * Configuration listener class that is intended to complete catch-up future during recovery when configuration
+     * is up-to-date.
+     */
+    private class ConfigurationCatchUpListener implements ConfigurationStorageRevisionListener {
+        /** Revision to catch up. */
+        private long cfgRevision;
+
+        /** Catch-up future. */
+        private final CompletableFuture<Void> catchUpFuture;
+
+        /**
+         * Constructor.
+         *
+         * @param cfgRevision Revision to catch up.
+         * @param catchUpFuture Catch-up future.
+         */
+        public ConfigurationCatchUpListener(long cfgRevision, CompletableFuture<Void> catchUpFuture) {
+            this.cfgRevision = cfgRevision;
+            this.catchUpFuture = catchUpFuture;
+        }
+
+        /** {@inheritDoc} */
+        @Override public CompletableFuture<?> onUpdate(long cfgUpdateRevision) {
+            assert cfgRevision >= cfgUpdateRevision : IgniteStringFormatter.format(
```

Comment:
Seems that currently there's a race between assertion (listener registration) and initial retrieval of last revision. Please check my comment above.